psec/pinblock.py

```diff
@@ -164,6 +171,157 @@ def encode_pinblock_iso_3(pin: str, pan: str) -> bytes:
+        - Fill digits. Each digit is a 4 bit hex value set to A.
+        - Random pad character. A 4 bit hex value in the range from 0 to F.
+
+    Parameters
+    ----------
+    pin :   str
+            ASCII Personal Identification Number.
+    det_pad (opt) : bytes
+                    Optional binary 8-byte deterministic padding to be used instead of random padding
+                    supporting deterministic test cases.
+
+    Returns
+    -------
+    Plain text PIN field : bytes
+        Binary 16-byte PIN field block.
+
+    Raises
+    ------
+    ValueError
+        PIN must be between 4 and 12 digits long.
+        Padding must be 8 bytes long.
+
+    Examples
+    --------
+    >>> from psec.pinblock import encode_pin_field_iso_4
+    >>> det_pad = bytes.fromhex("548ED7FD65495950")
+    >>> encode_pin_field_iso_4("1234", det_pad).hex().upper()
+    '441234AAAAAAAAAA548ED7FD65495950'
+    """
+
+    if len(pin) < 4 or len(pin) > 12 or not _tools.ascii_numeric(pin):
+        raise ValueError("PIN must be between 4 and 12 digits long")
+
+    if det_pad is None:
+        random_pad = "".join(_secrets.choice("0123456789ABCDEF") for _ in range(16))
+    else:
+        if len(det_pad) != 8:
+            raise ValueError("Padding must be 8 bytes long.")
+        random_pad = det_pad.hex().upper()
+
+    pin_len_hex = len(pin).to_bytes(1, "big").hex()[1]  # Only the low nibble is relevant, values 4 - C.
+    pinblock_str = "4" + pin_len_hex + pin + "A" * (14 - len(pin)) + random_pad
+    return _binascii.a2b_hex(pinblock_str)
+
+
+def encode_pan_field_iso_4(pan: str) -> bytes:
+    r"""Encode ISO 9564 PIN block format 4 plain text primary account number (PAN) field.
+    ISO format 4 plain text primary account number field is a 16 byte value that consits of
+
+        - PAN length. A 4 bit hex value in the range from 0 to 7 indicate a PAN length of 12 plus the value of the field
+                      (ranging from dec. 12 to 19). If the PAN is less than 12 digits, the digits are right justified and
+                      padded to the left with zeros and PAN length is set to 0.
+        - PAN digits. Each digit is a 4 bit hex value in the range from 0 to 9.
+        - Pad digits. A 4 bit hex value set to 0.
+
+    Parameters
+    ----------
+    pan :   str
+            ASCII Personal Account Number.
+
+    Returns
+    -------
+    Plain text PAN field :  bytes
+                            Binary 16-byte PAN field.
+
+    Raises
+    ------
+    ValueError
+        PAN must be between 1 and 19 digits long.
+
+    Examples
+    --------
+    >>> from psec.pinblock import encode_pan_field_iso_4
+    >>> encode_pan_field_iso_4("112233445566778899").hex().upper()
+    '61122334455667788990000000000000'
+    """
+
+    if len(pan) < 1 or len(pan) > 19 or not _tools.ascii_numeric(pan):
+        raise ValueError("PAN must be between 1 and 19 digits long.")
+
+    pan_len_field = "0" if len(pan) < 12 else str(len(pan) - 12)
```

Comment:
nit: if you pad the pan first then you don't need to have an if/else here.

Reply:
I made a one liner to construct the pan_field. I hope this is ok.